src/main/java/io/vertx/core/http/impl/Http2ServerResponseImpl.java

```diff
@@ -0,0 +1,629 @@
+      checkEnded();
+      if (end) {
+        handleEnded(false);
+      }
+      boolean hasBody = chunk != null;
+      boolean sent = checkSendHeaders(end && !hasBody && trailers == null);
+      if (hasBody || (!sent && end)) {
+        if (chunk == null) {
+          chunk = Unpooled.EMPTY_BUFFER;
+        }
+        int len = chunk.readableBytes();
+        stream.writeData(chunk, end && trailers == null);
+        bytesWritten += len;
+      }
+      if (end && trailers != null) {
+        stream.writeHeaders(trailers, true);
+      }
+      if (end && bodyEndHandler != null) {
+        bodyEndHandler.handle(null);
+      }
+    }
+  }
+
+  @Override
+  public HttpServerResponse writeFrame(int type, int flags, Buffer payload) {
+    synchronized (conn) {
+      checkEnded();
+      checkSendHeaders(false);
+      stream.writeFrame(type, flags, payload.getByteBuf());
+      ctx.flush();
+      return this;
+    }
+  }
+
+  private void checkEnded() {
+    if (ended) {
+      throw new IllegalStateException("Response has already been written");
+    }
+  }
+
+  private boolean handleEnded(boolean failed) {
+    if (!ended) {
+      ended = true;
+      if (metric != null) {
+        // Null in case of push response : handle this case
+        if (failed) {
+          conn.metrics().requestReset(metric);
+        } else {
+          conn.reportBytesWritten(bytesWritten);
+          conn.metrics().responseEnd(metric, this);
+        }
+      }
+      return true;
+    }
+    return false;
+  }
+
+  void writabilityChanged() {
+    if (!ended && !writeQueueFull() && drainHandler != null) {
+      drainHandler.handle(null);
+    }
+  }
+
+  @Override
+  public boolean writeQueueFull() {
+    synchronized (conn) {
+      checkEnded();
+      return stream.isNotWritable();
+    }
+  }
+
+  @Override
+  public HttpServerResponse setWriteQueueMaxSize(int maxSize) {
+    synchronized (conn) {
+      checkEnded();
+      // It does not seem to be possible to configure this at the moment
+    }
+    return this;
+  }
+
+  @Override
+  public HttpServerResponse drainHandler(Handler<Void> handler) {
+    synchronized (conn) {
+      checkEnded();
+      drainHandler = handler;
+      return this;
+    }
+  }
+
+  @Override
+  public HttpServerResponse sendFile(String filename, long offset, long length) {
+    return sendFile(filename, offset, length, null);
+  }
+
+  @Override
+  public HttpServerResponse sendFile(String filename, long offset, long length, Handler<AsyncResult<Void>> resultHandler) {
+    synchronized (conn) {
+      checkEnded();
+
+      Context resultCtx = resultHandler != null ? stream.vertx.getOrCreateContext() : null;
+
+      File file = stream.vertx.resolveFile(filename);
+      if (!file.exists()) {
+        if (resultHandler != null) {
+          resultCtx.runOnContext((v) -> resultHandler.handle(Future.failedFuture(new FileNotFoundException())));
+        } else {
+          // log.error("File not found: " + filename);
```

Comment:
I need to figure out.

> On Apr 6, 2016, at 11:00 AM, Clement Escoffier <notifications@github.com> wrote:
> 
> In src/main/java/io/vertx/core/http/impl/Http2ServerResponseImpl.java <https://github.com/eclipse/vert.x/pull/1353#discussion_r58672949>:
> 
> > +    return sendFile(filename, offset, length, null);
> > +  }
> > +
> > +  @Override
> > +  public HttpServerResponse sendFile(String filename, long offset, long length, Handler<AsyncResult<Void>> resultHandler) {
> > +    synchronized (conn) {
> > +      checkEnded();
> > +
> > +      Context resultCtx = resultHandler != null ? stream.vertx.getOrCreateContext() : null;
> > +
> > +      File file = stream.vertx.resolveFile(filename);
> > +      if (!file.exists()) {
> > +        if (resultHandler != null) {
> > +          resultCtx.runOnContext((v) -> resultHandler.handle(Future.failedFuture(new FileNotFoundException())));
> > +        } else {
> > +          // log.error("File not found: " + filename);
> should we do something here ? If not maybe a comment explaining why.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/eclipse/vert.x/pull/1353/files/bf875485313cf1f6ae199e2ab525539ed5676a2e#r58672949>